control-plane/agents/core/src/volume/specs.rs

```diff
@@ -464,9 +436,416 @@ impl ResourceSpecsLocked {
             })
         }
     }
+
+    pub(crate) async fn share_volume(
+        &self,
+        registry: &Registry,
+        request: &ShareVolume,
+    ) -> Result<String, SvcError> {
+        let volume_spec = self
+            .get_volume(&request.uuid)
+            .await
+            .context(errors::VolumeNotFound {
+                vol_id: request.uuid.to_string(),
+            })?;
+        let status = registry.get_volume_status(&request.uuid).await?;
+
+        let spec_clone = {
+            let mut spec = volume_spec.lock().await;
+            if spec.pending_op() {
+                return Err(SvcError::StoreSave {
+                    kind: ResourceKind::Volume,
+                    id: request.uuid.to_string(),
+                });
+            } else if spec.updating {
+                return Err(SvcError::Conflict {});
+            } else if !spec.state.created() {
+                return Err(SvcError::VolumeNotFound {
+                    vol_id: request.uuid.to_string(),
+                });
+            } else {
+                // validate the operation itself against the spec and status
+                if spec.protocol != status.protocol {
+                    return Err(SvcError::NotReady {
+                        kind: ResourceKind::Volume,
+                        id: request.uuid.to_string(),
+                    });
+                } else if request.protocol == spec.protocol || spec.protocol.shared() {
+                    return Err(SvcError::AlreadyShared {
+                        kind: ResourceKind::Volume,
+                        id: spec.uuid.to_string(),
+                        share: spec.protocol.to_string(),
+                    });
+                }
+                if status.children.len() != 1 {
+                    return Err(SvcError::NotReady {
```

Comment:
I wonder if we should have an ANA not supported error?

Reply:
Not sure, this is checking the status part of the volume. We can end up with N nexuses in we loose connectivity to a node, then decide to create the nexus on another node and then the first node is reachable again and reports the old nexus. Once we have the persistent reserves we can probably ignore this nexus as it won't be able to write to the children anymore.

Reply:
Ok